### DISCUSSION TITLE
Is there a way to distinguish a Combobox value that was selected from the popover vs just typed in?

### DISCUSSION THREAD

**Discussion by candidia** [ariakit/ariakit#2525] [Questions] [answered]
I'd like to only allow submitting a form if the value in the combobox was actually selected from the popover items, not just anything typed into it. Is there a built-in way to distinguish this?

**Comment by diegohaz** [comment]
You have a couple of options to handle the scenario you described:

1. If you only allow values that were explicitly selected from the list of options, you can maintain a separate state to track the last selected value. This can be achieved by using the `onClick` prop on the `ComboboxItem` component, similar to how you would handle a list of native buttons.

2. If you allow any value in the list, regardless of whether it was selected from the options or typed manually, you can compare the input value against the list of available values. Again, this approach is not specific to Ariakit. You can check if the input value matches any of the values in the list, and if it does, you can consider it a valid value.

However, what you're describing aligns with a [Select with Combobox](https://ariakit.org/examples/select-combobox) where the primary input allows selection from the available options while still providing search functionality. This approach can offer a better user experience as it makes it clear to users that they need to choose a value from the provided list. With a free-text combobox, there is a possibility that users may become frustrated when they realize that the value they entered cannot be used.

**Reply by candidia** [reply]
My case is a little more complex because my combobox fetches different items from server for every keystroke. There is no static single set of all possible items on the client. Seems like I'd also need to construct a set of items from all items returned across fetches.

I like how you distinguished the two cases and think the second is the nicest, but it seems extra complex when combining with this case. I wonder if you would consider just supporting the first case natively by providing something like  `combobox.useState('wasValueSelected')`? I feel like a native way to distinguish between a value that was typed and selected would be cleaner than writing custom logic each time.

**Comment by michaelsawyers** [comment, marked as answer]
This is now made easy with the addition of `selectedValue` to the ComboboxStore, as of 0.3.8. 

`value` is the value of the input. So if user types 'foo' and does not select anything, then`value` is 'foo' and `selectedValue` is empty. 
But if user selects 'food' item after typing 'foo', the `value` changes from `foo` to `food` and the `selectedValue` changes from empty to 'food'.

You can read up on this here: https://github.com/ariakit/ariakit/releases/tag/@ariakit/react@0.3.8
Or see the update multi-select example here (which is much simpler than previous versions 😄): https://ariakit.org/examples/combobox-multiple
